src/main/java/net/fabricmc/meta/web/EndpointsV2.java

```diff
@@ -34,145 +34,170 @@
+        }
+        return new LoaderInfoV2(loader, mappings).populateMeta();
+    }
 
-		if (loader == null) {
-			context.status(400);
-			return "no loader version found for " + gameVersion;
-		}
-		if (mappings == null) {
-			context.status(400);
-			return "no mappings version found for " + gameVersion;
-		}
-		return new LoaderInfoV2(loader, mappings).populateMeta();
-	}
+    private static List<?> getLoaderInfoAll(Context context) {
+        if (!context.pathParamMap().containsKey("game_version")) {
+            return null;
+        }
+        String gameVersion = context.pathParam("game_version");
 
-	private static List<?> getLoaderInfoAll(Context context) {
-		if (!context.pathParamMap().containsKey("game_version")) {
-			return null;
-		}
-		String gameVersion = context.pathParam("game_version");
+        MavenVersion mappings = FabricMeta.database.intermediary.stream()
+                .filter(t -> t.test(gameVersion))
+                .findFirst().orElse(null);
 
-		MavenVersion mappings = FabricMeta.database.intermediary.stream()
-			.filter(t -> t.test(gameVersion))
-			.findFirst().orElse(null);
+        if (mappings == null) {
+            return Collections.emptyList();
+        }
 
-		if(mappings == null){
-			return Collections.emptyList();
-		}
+        List<LoaderInfoV2> infoList = new ArrayList<>();
 
-		List<LoaderInfoV2> infoList = new ArrayList<>();
+        for (MavenBuildVersion loader : FabricMeta.database.getLoader()) {
+            infoList.add(new LoaderInfoV2(loader, mappings).populateMeta());
+        }
+        return infoList;
+    }
 
-		for(MavenBuildVersion loader : FabricMeta.database.getLoader()){
-			infoList.add(new LoaderInfoV2(loader, mappings).populateMeta());
+    private static Object getRecommendedVersions(Context context) {
+        if (!context.pathParamMap().containsKey("game_version")) {
+            return null;
+        }
+        String gameVersion = context.pathParam("game_version");
+
+        List<MavenBuildGameVersion> yarn = FabricMeta.database.mappings.stream()
+                .filter(t -> t.test(gameVersion))
+                .collect(Collectors.toList());
+
+		List<MavenBuildVersion> loader = FabricMeta.database.getLoader();
+
+		if (yarn.isEmpty()) {
+			return null;
 		}
-		return infoList;
-	}
-
-	private static <T extends BaseVersion> List<BaseVersion> compatibleGameVersions(List<T> list, Function<T, String> gameVersionSupplier, Function<T, BaseVersion> baseVersionSupplier){
-		List<BaseVersion> versions = new ArrayList<>();
-		Predicate<String> contains = s -> versions.stream().anyMatch(baseVersion -> baseVersion.getVersion().equals(s));
 
-		for(T entry : list){
-			if (!contains.test(gameVersionSupplier.apply(entry))){
-				versions.add(baseVersionSupplier.apply(entry));
-			}
+		if (loader.isEmpty()) {
+			return null;
 		}
 
-		return versions;
-	}
+        return new RecommendedVersions(gameVersion, yarn.get(0), loader.get(0));
```

Comment:
It relies on the fact that the latest versions of yarn and loader are always first in the JSON - is this reliable?